torch/csrc/jit/passes/mkldnn_rewrite.cpp

```diff
@@ -0,0 +1,276 @@
+        %res = aten::${op}(%conv2d_res)
+        return (%res))");
+
+  auto conv_op_fused_rstring = at::jit::CodeTemplate(R"(
+    graph(%input, %weight, %bias, %stride:int[], %padding:int[],
+          %dilation:int[], %groups:int, %input_size:int[], %dummy_attr:str):
+        %attr: str = prim::Constant[value="${op_attr}"]()
+        %packed_weight_bias : __torch__.torch.classes.mkldnn.ConvOpContext = mkldnn_prepacked::conv2d_prepack(
+            %weight, %bias, %stride, %padding, %dilation, %groups,
+            %input_size, %attr)
+        %res = mkldnn_prepacked::conv2d_run(%input, %packed_weight_bias)
+        return (%res))");
+
+  for (auto const& it : mkldnn::fusion_attr_map) {
+    std::string op = it.first;
+    if (op == std::string("none")) {
+      continue;
+    }
+
+    at::jit::TemplateEnv env;
+    env.s("op", op);
+
+    at::jit::TemplateEnv env_fused;
+    env_fused.s("op_attr", op);
+
+    SubgraphRewriter rewriter;
+    rewriter.RegisterRewritePattern(
+        conv_op_rstring.format(env), conv_op_fused_rstring.format(env_fused));
+
+    auto filters = it.second.filters;
+    rewriter.runOnGraph(graph, filters);
+  }
+}
+
+void PrePackingOpsFolder(Block* b) {
+  auto is_foldable_op = [](const Node* n) -> bool {
+    return (
+        n->kind() ==
+        Symbol::fromQualString("mkldnn_prepacked::conv2d_prepack"));
+  };
+
+  std::unordered_set<Node*> nodes_to_delete;
+  for (Node* n : b->nodes()) {
+    for (Block* block : n->blocks()) {
+      PrePackingOpsFolder(block);
+    }
+    if (is_foldable_op(n)) {
+      auto optional_outputs = torch::jit::runNodeIfInputsAreConstant(n);
+      if (optional_outputs) {
+        auto outputs = optional_outputs.value();
+        TORCH_CHECK(outputs.size() == 1, "Prepack ops have single output");
+        Value* prepack_op_value = n->output(0);
+        auto graph = n->owningGraph();
+        WithInsertPoint ins(prepack_op_value->node());
+        auto weak_class_obj =
+            outputs[0].toObject()->copy_to_weak_compilation_ref();
+        Value* packed_weight = graph->insertConstant(weak_class_obj)
+                                   ->setType(n->output(0)->type());
+        prepack_op_value->replaceAllUsesWith(packed_weight);
+        nodes_to_delete.insert(n);
+      }
+    }
+  }
+  for (auto n : nodes_to_delete) {
+    n->removeAllInputs();
+  }
+  for (auto n : nodes_to_delete) {
+    n->destroy();
+  }
+}
+
+void FoldPrePackingOps(std::shared_ptr<Graph>& graph) {
+  PrePackingOpsFolder(graph->block());
+}
+
+void FuseConvWithEltwise(std::shared_ptr<Graph>& graph) {
+  GRAPH_DEBUG(
+      "Before insertMkldnnPrePackedOps. Beginning of FuseConvWithEltwise\n",
+      *graph);
+  insertMkldnnPrePackedOps(graph);
+  GRAPH_DEBUG(
+      "After insertMkldnnPrePackedOps, before FuseReluWithPackedOps\n", *graph);
+  FuseReluWithPackedOps(graph);
+  GRAPH_DEBUG(
+      "After FuseReluWithPackedOps, before FoldPrePackingOps\n", *graph);
+  FoldPrePackingOps(graph);
```

Comment:
Can we use the standard ConstantPropagation pass here? Does it not work for some reason?

Reply:
Inside the ConstantPropagation pass, when calling `runNodeIfInputsAreConstant`, the 3rd parameter `db` is not provided thus the default value `nullptr` will be used:
https://github.com/pytorch/pytorch/blob/9e806619ccbd211b2501fd380ec7d1bd398ec67c/torch/csrc/jit/passes/constant_propagation.cpp#L180-L183
Default value of `db` is `nullptr`:
https://github.com/pytorch/pytorch/blob/9e806619ccbd211b2501fd380ec7d1bd398ec67c/torch/csrc/jit/passes/constant_propagation.h#L26-L29

However for Custom class, if `!db`, the constant propagation won't be functional:
https://github.com/pytorch/pytorch/blob/9e806619ccbd211b2501fd380ec7d1bd398ec67c/torch/csrc/jit/passes/constant_propagation.cpp#L107-L113

In our case, the prepacked OP that we're trying to do the constant propagation is a custom class thus it does not work.

Is this by design or an issue that need to be fixed in the standard ConstantPropagation pass? If it is an issue, I can submit another PR to fix it.

Reply:
I think we should just pass `aliasDb_` to `runNodeIfInputsAreConstant`. @eellison, do you know why we're not doing it now?